tasks/fine_tuning_sbert/src/loops.py

```diff
@@ -61,79 +68,168 @@ def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor
             return features, output
 
 
+def train(config=None):
+    """
+    Find the optimal SBERT model by doing a hyperparameter search over random seeds, dev percentage, and different types of SBERT models
+    """
+
+    # this will write to the same project every time
+    wandb.init(config=config, magic=True)
+
+    config = wandb.config
+
+    print(
+        f"Grid Search Fine tuning parameters:\n{config}")
+
+    label2int = dict(zip(label_names, range(len(label_names))))
+
+    model_deets = f"{config.eval_classifier}_model={config.model_name}_test-perc={config.dev_perc}_seed={config.seeds}"
+
+    wandb.run.notes = model_deets
+
+    X_train, X_dev, y_train, y_dev = train_test_split(train_sents, train_labels, test_size=config.dev_perc,
+                                                      stratify=train_labels, random_state=100)
+
+    # Load data samples into batches
+    train_batch_size = 16
+    train_samples = build_data_samples(X_train, label2int, y_train)
+    dev_samples = build_data_samples(X_dev, label2int, y_dev)
+
+    # Train set config
+    model = EarlyStoppingSentenceTransformer(config.model_name)
+    train_dataset = SentencesDataset(train_samples, model=model)
+    train_dataloader = DataLoader(
+        train_dataset, shuffle=True, batch_size=train_batch_size)
+
+    # Dev set config
+    dev_dataset = SentencesDataset(dev_samples, model=model)
+    dev_dataloader = DataLoader(
+        dev_dataset, shuffle=True, batch_size=train_batch_size)
+
+    # Define the way the loss is computed
+    classifier = SoftmaxClassifier(model=model,
+                                   sentence_embedding_dimension=model.get_sentence_embedding_dimension(),
+                                   num_labels=len(label2int))
+    warmup_steps = math.ceil(
+        len(train_dataset) * config.max_num_epochs / train_batch_size * 0.1)  # 10% of train data for warm-up
+
+    set_seeds(config.seeds)
+
+    # Train the model
+    start = time.time()
+    dev_evaluator = CustomLabelAccuracyEvaluator(dataloader=dev_dataloader, softmax_model=classifier,
+                                                 name='lae-dev', label_names=label_names,
+                                                 model_hyper_params={'model_name': config.model_name, 'dev_perc': config.dev_perc, 'seed': config.seeds})
+
+    model.fit(train_objectives=[(train_dataloader, classifier)],
+              evaluator=dev_evaluator,
+              epochs=config.max_num_epochs,
+              evaluation_steps=1000,
+              warmup_steps=warmup_steps,
+              output_path=config.output_path,
```

Comment:
Why is the output path part of the config? Can we leave it out so we don't optimize for it :D?

Reply:
I included output_path here so it didn't have to be another global variable :) 
We could make one constant output_path for all experiments where the model would be saved but that would be problematic if more than one person saves a model on the google drive folder at the same time. I think it needs to be an input from the notebook for now.
It's not optimized over because we only pass in 1 value to the sweep config for output_path 😁

Reply:
Gotcha, thanks!